remotes/workqueue.go

```diff
@@ -0,0 +1,83 @@
+package remotes
+
+import (
+	"context"
+
+	"golang.org/x/sync/errgroup"
+)
+
+type workItem struct {
+	do   func(ctx context.Context) error
+	done chan struct{}
+	deps []chan struct{}
+}
+
+func newWorkItem(do func(ctx context.Context) error, deps ...chan struct{}) *workItem {
+	return &workItem{
+		do:   do,
+		done: make(chan struct{}),
+		deps: deps,
+	}
+}
+
+func (wi *workItem) process(ctx context.Context) error {
+	defer close(wi.done)
+	for _, d := range wi.deps {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-d:
+		}
+	}
+	select {
+	case <-ctx.Done():
+		return nil
+	default:
+		return wi.do(ctx)
+	}
+}
+
+type workQueue struct {
+	workGroup *errgroup.Group
+	todoList  chan *workItem
+	ctx       context.Context
+}
+
+//nolint: unparam
+func newWorkQueue(ctx context.Context, workerCount, todoBuffer int) *workQueue {
+	todoList := make(chan *workItem, todoBuffer)
+	workGroup, ctx := errgroup.WithContext(ctx)
+	for i := 0; i < workerCount; i++ {
+		workGroup.Go(func() error {
+			for {
+				wi, ok := <-todoList
+				if !ok {
+					return nil
+				}
+				if err := wi.process(ctx); err != nil {
+					return err
+				}
+			}
+		})
+	}
+	return &workQueue{
+		todoList:  todoList,
+		workGroup: workGroup,
+		ctx:       ctx,
+	}
+}
+
+func (wq *workQueue) enqueue(do func(ctx context.Context) error, deps ...chan struct{}) chan struct{} {
+	wi := newWorkItem(do, deps...)
+	select {
+	case <-wq.ctx.Done():
+		return doneCh
+	case wq.todoList <- wi:
+	}
+	return wi.done
+}
+
+func (wq *workQueue) stopAndWait() error {
+	close(wq.todoList)
+	return wq.workGroup.Wait()
```

Comment:
Since each `workItem`'s `do` takes a `ctx` I think `newWorkQueue` should call `WithCancel` on the `ctx` it is given and stash the resulting  `cancel` func into the `wq`, then here you should call it. This will mean that any `workItem.do` which uses the context properly will, where possible, stop without waiting for completion.

Reply:
No, stopAndWait, is for notifying the queue that there wont be more work after this point, and to wait for the workers to have handled every work item.
If we do cancel at this point and there is still work items pending in the queue, we will cancel there execution (and we will have workgroup.Wait returning an error). Perhaps the naming is wrong.

(to trigger cancellation, one has only to cancel the context passed in newWorkQueue).

Reply:
Ah, yes, the name `stop` is a bit confusing then. I think just `wait` would be fine, perhaps with a function comment saying "you mustn't enqueue anything more after this point" (if that's true).

Reply:
:+1